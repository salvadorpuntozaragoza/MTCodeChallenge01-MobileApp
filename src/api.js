import { getItem } from "./asyncStoreManager";

const URL = 'http://10.0.2.2:3000/';

async function getAuthorizationToken() {
  const item = await getItem('session');
  if(item == null) {
    return item;
  }
  return item.token;
}

const fetchApi = async (url, method, params) => {
  try {
    const token = await getAuthorizationToken();
    console.log('Calling api with token: ', token);
    const response = await fetch(url, {
      method: method,
      headers: {
        'Content-Type': 'application/json',
        'access-token': token
      },
      body: (method === 'GET') ? null : JSON.stringify(params)
    });
    console.log(response);
    const json = await response.json();
    return json;
  } catch (error) {
    console.log('Error catched: ', error);
    return error;
  }
}

export const get = async (path, body = {}) => {
  const response = await fetchApi(`${URL}${path}`, 'GET', body);
  return response;
}

export const post = async (path, body = {}) => {
  const response = await fetchApi(`${URL}${path}`, 'POST', body);
  return response;
}

export const put = async (path, body = {}) => {
  const response = await fetchApi(`${URL}${path}`, 'PATCH', body);
  return response;
}

export const del = async (path, body = {}) => {
  const response = await fetchApi(`${URL}${path}`, 'DELETE', body);
  return response;
}