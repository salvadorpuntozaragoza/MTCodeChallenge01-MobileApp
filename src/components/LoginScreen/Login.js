import React, { useState } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import { Button, Icon, Input } from 'react-native-elements';
import { ScrollView, TouchableOpacity } from 'react-native-gesture-handler';
import { useDispatch, useSelector } from 'react-redux';
import { COLORS } from '../../assets/colors';
import { logIn, signUp } from '../../redux/actions';
import { validateEmail } from '../../utils';

const styles = StyleSheet.create({
  buttonStyle: {
    borderRadius: 20,
    marginHorizontal: 15,
  },
  button: {
    backgroundColor: COLORS.primaryColor
  },
  container: {
    alignSelf: 'stretch',
    backgroundColor: COLORS.secondaryColorDarker
  },
  inputStyle: {
    borderBottomColor: COLORS.primaryColor,
  },
  header: {
    justifyContent: 'center',
    alignItems: "flex-start",
    paddingHorizontal: 10,
    width: '100%',
    height: 50,
  },
  errorMessage: {
    marginTop: 15,
    alignSelf: 'center',
    color: COLORS.primaryColor,
  },
  noAccountText: {
    color: COLORS.white,
    alignSelf: 'center',
    fontSize: 18,
    marginTop: 15,
  },
  createAccountText: {
    color: COLORS.primaryColorDarker,
    alignSelf: 'center',
    fontSize: 18,
    textDecorationLine: "underline",
    marginTop: 15,
  },
  headerTitle: {
    color: COLORS.white,
    alignSelf: 'center',
    fontSize: 25,
    fontWeight: "bold",
  }
});

const Login = ({ navigation, store, dispatch }) => {
  // const store = useSelector(({ loginReducer, sessionReducer }) => ({loginReducer, sessionReducer}));
  // const dispatch = useDispatch();

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  const [invalidEmail, setInvalidEmail] = useState(false);
  const [invalidPassword, setInvalidPassword] = useState(false);

  console.log("Session reducer: ", store.sessionReducer);
  console.log("Login reducer: ", store.loginReducer);

  const handleEmailChange = (value) => {
    setEmail(value);
    setInvalidEmail(!validateEmail(value));
  }

  const handlePasswordChange = (value) => {
    setPassword(value);
    setInvalidPassword(value.length < 6);
  }

  const isSubmitButtonDisabled = () => (
    password.length < 6
    || !validateEmail(email)
  );

  const handleOnPress = () => {
    dispatch(logIn({
      data: {
        email: email,
        password: password,
      },
      navigation: navigation,
    }));
  }

  const handleCreateAccount = () => {
    navigation.navigate('Register');
  }

  return (
    <ScrollView style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.headerTitle}>Login</Text>
      </View>

      <Input
        value={email}
        onChangeText={handleEmailChange}
        errorMessage={ invalidEmail ? "Invalid email" : "" }
        placeholder="Email"
        placeholderTextColor='white'
        leftIcon={{ type: 'font-awesome', name: 'envelope', color: 'white', size: 16 }}
        inputContainerStyle={styles.inputStyle}
        inputStyle={{ color: "white" }}
      />

      <Input
        value={password}
        secureTextEntry
        onChangeText={handlePasswordChange}
        errorMessage={ invalidPassword ? "Password must be at least 6 characters long" : "" }
        placeholder="Password"
        placeholderTextColor='white'
        leftIcon={{ type: 'font-awesome', name: 'lock', color: 'white' }}
        inputContainerStyle={styles.inputStyle}
        inputStyle={{ color: "white" }}
      />

      <Button
        disabled={isSubmitButtonDisabled()}
        disabledStyle={{ backgroundColor: '#cccccc' }}
        containerStyle={styles.buttonStyle}
        buttonStyle={styles.button}
        loading={store.loginReducer.signinIn}
        title="Sign up"
        onPress={handleOnPress}
        raised
      />

      <Text style={styles.errorMessage}>
        {store.loginReducer.errorMessage}
      </Text>

      <Text style={styles.noAccountText}>Don't have an account?</Text>
      
      <TouchableOpacity onPress={handleCreateAccount}>
        <Text style={styles.createAccountText}>Create one</Text>
      </TouchableOpacity>

    </ScrollView>
  )
}

export { Login }