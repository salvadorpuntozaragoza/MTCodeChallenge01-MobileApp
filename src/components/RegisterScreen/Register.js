import React, { useState } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import { Button, Icon, Input } from 'react-native-elements';
import { ScrollView } from 'react-native-gesture-handler';
import { useDispatch, useSelector } from 'react-redux';
import { COLORS } from '../../assets/colors';
import { signUp } from '../../redux/actions';
import { validateEmail } from '../../utils';

const styles = StyleSheet.create({
  buttonStyle: {
    borderRadius: 20,
    marginHorizontal: 15,
  },
  button: {
    backgroundColor: COLORS.primaryColor
  },
  container: {
    alignSelf: 'stretch',
    backgroundColor: COLORS.secondaryColorDarker
  },
  inputStyle: {
    borderBottomColor: COLORS.primaryColor,
  },
  header: {
    justifyContent: 'center',
    alignItems: "flex-start",
    paddingHorizontal: 10,
    width: '100%',
    height: 50,
  },
  errorMessage: {
    color: COLORS.primaryColor,
  }
});

const Register = ({ navigation }) => {
  const store = useSelector(({ registerReducer }) => registerReducer);
  const dispatch = useDispatch();

  const [fullname, setFullname] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmPass, setConfirmPass] = useState('');
  const [disabled, setDisabled] = useState(true);

  const [invalidFullname, setInvalidFullname] = useState(false);
  const [invalidEmail, setInvalidEmail] = useState(false);
  const [invalidPassword, setInvalidPassword] = useState(false);
  const [invalidConfirmPass, setInvalidConfirmPass] = useState(false);

  const handleFullnameChange = (value) => {
    setFullname(value);
    setInvalidFullname(value.length < 3);
  }

  const handleEmailChange = (value) => {
    setEmail(value);
    setInvalidEmail(!validateEmail(value));
  }

  const handlePasswordChange = (value) => {
    setPassword(value);
    setInvalidPassword(value.length < 6);
  }

  const handleConfirmPasswordChange = (value) => {
    setConfirmPass(value);
    setInvalidConfirmPass(value !== password);
  }

  const isSubmitButtonDisabled = () => (
    fullname.length < 3
    || password.length < 6
    || (confirmPass !== password)
    || !validateEmail(email)
  );

  const handleOnPress = () => {
    dispatch(signUp({
      data: {
        name: fullname,
        email: email,
        password: password,
      },
      navigation: navigation,
    }));
  }

  return (
    <ScrollView style={styles.container}>
      <View style={styles.header}>
        <Icon
          type="font-awesome"
          name="arrow-left"
          color={COLORS.white}
          size={20}
          onPress={() => navigation.goBack()}
        />
      </View>
      <Input 
        value={fullname}
        placeholder="Full name"
        placeholderTextColor='white'
        errorMessage={ invalidFullname ? "This field is required" : "" }
        onChangeText={handleFullnameChange}
        leftIcon={{ type: 'font-awesome', name: 'user', color: COLORS.white }}
        inputContainerStyle={styles.inputStyle}
        inputStyle={{ color: "white" }}
      />
      <Input
        value={email}
        onChangeText={handleEmailChange}
        errorMessage={ invalidEmail ? "Invalid email" : "" }
        placeholder="Email"
        placeholderTextColor='white'
        leftIcon={{ type: 'font-awesome', name: 'envelope', color: COLORS.white, size: 16 }}
        inputContainerStyle={styles.inputStyle}
        inputStyle={{ color: "white" }}
      />
      <Input
        value={password}
        secureTextEntry
        onChangeText={handlePasswordChange}
        errorMessage={ invalidPassword ? "Password must be at least 6 characters long" : "" }
        placeholder="Password"
        placeholderTextColor='white'
        leftIcon={{ type: 'font-awesome', name: 'lock', color: COLORS.white }}
        inputContainerStyle={styles.inputStyle}
        inputStyle={{ color: "white" }}
      />
      <Input
        value={confirmPass}
        secureTextEntry
        onChangeText={handleConfirmPasswordChange}
        errorMessage={ invalidConfirmPass ? "The password doesn't match" : "" }
        placeholder="Confirm password"
        placeholderTextColor='white'
        leftIcon={{ type: 'font-awesome', name: 'lock', color: COLORS.white }}
        inputContainerStyle={styles.inputStyle}
        inputStyle={{ color: "white" }}
      />
      <Button
        disabled={isSubmitButtonDisabled()}
        disabledStyle={{ backgroundColor: '#cccccc' }}
        containerStyle={styles.buttonStyle}
        buttonStyle={styles.button}
        loading={store.signingUp}
        title="Sign up"
        onPress={handleOnPress}
        raised
      />
      <Text style={styles.errorMessage}>
        {store.message}
      </Text>
    </ScrollView>
  )
}

export { Register }
