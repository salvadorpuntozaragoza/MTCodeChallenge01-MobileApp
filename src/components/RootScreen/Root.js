import React, { useState } from 'react';
import { ScrollView } from 'react-native-gesture-handler';
import { View, StyleSheet, Image, Modal, Text, ActivityIndicator } from 'react-native';
import { Button } from 'react-native-elements';
import { useDispatch } from 'react-redux';
import {
  heightPercentageToDP as height,
  widthPercentageToDP as width,
} from 'react-native-responsive-screen'
import { COLORS } from '../../assets/colors';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignSelf: "stretch",
    backgroundColor: COLORS.secondaryColorDarker,
  },
  imageContainer: {
    flex: 2,
    alignItems: 'center',
    alignSelf: 'center',
    height: height(100),
    width: width(100),
  },
  buttonContainer: {
    alignSelf: 'center',
    flex: 1,
    height: height(30),
    justifyContent: 'center',
    paddingHorizontal: width(5),
    width: width(85),
  },
  buttonStyle: {
    borderRadius: 20
  },
  button: {
    backgroundColor: COLORS.primaryColor
  },
  textStyle: {
    alignSelf: 'center',
    color: COLORS.white,
    fontSize: 12,
    marginVertical: 10,
  },
});

const Root = ({ navigation }) => {
  const [isVisible, setIsVisible] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(false);
  const [success, setSuccess] = useState(false);

  return (
    <View style={styles.container}>
      <View style = { styles.imageContainer }>
        <Image 
          source={require('../../assets/images/logo.png')}
          width={60}
          height={60}
          resizeMode="center"
        />
      </View>
      <View style = { styles.buttonContainer }>
        <Button
          containerStyle={styles.buttonStyle}
          buttonStyle={styles.button}
          title="Log in"
          onPress={() => navigation.navigate('Login')}
          raised
        />
        <Text style={styles.textStyle}>
          Or
        </Text>
        <Button
          containerStyle={styles.buttonStyle}
          buttonStyle={styles.button}
          title="Sign up"
          onPress={() => navigation.navigate('Register')}
          raised
        />
      </View>
    </View>
  )
}

export {Root}
