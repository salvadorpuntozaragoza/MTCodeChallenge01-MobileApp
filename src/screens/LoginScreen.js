import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { ContainerView } from '../components/ContainerView/ContainerView';
import { Login } from '../components/LoginScreen/Login'

const LoginScreen = ({ navigation }) => {
  const store = useSelector(({ loginReducer, sessionReducer }) => ({loginReducer, sessionReducer}));
  const dispatch = useDispatch();
  
  return (
    <ContainerView>
      <Login
        navigation={navigation}
        store={store}
        dispatch={dispatch}
      />
    </ContainerView>
  );
}

export { LoginScreen };