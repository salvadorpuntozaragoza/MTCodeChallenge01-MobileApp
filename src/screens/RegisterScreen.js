import React from 'react';
import { ContainerView } from '../components/ContainerView/ContainerView';
import { Register } from '../components/RegisterScreen/Register';

const RegisterScreen = ({ navigation }) => (
  <ContainerView>
    <Register navigation={navigation}/>
  </ContainerView>
);

export { RegisterScreen };
