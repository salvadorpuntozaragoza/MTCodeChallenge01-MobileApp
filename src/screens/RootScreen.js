import React from 'react';
import { ContainerView } from '../components/ContainerView/ContainerView';
import { Root } from '../components/RootScreen/Root';

const RootScreen = ({ navigation }) => (
  <ContainerView>
    <Root navigation={navigation}/>
  </ContainerView>
);

export { RootScreen };