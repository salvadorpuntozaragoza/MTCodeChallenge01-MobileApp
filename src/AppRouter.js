import React from 'react';
import { StyleSheet } from 'react-native';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs'

import { HomeScreen } from './screens/HomeScreen';
import { ProfileScreen } from './screens/ProfileScreen';
import { RootScreen } from './screens/RootScreen';
import { RegisterScreen } from './screens/RegisterScreen';
import { LoginScreen } from './screens/LoginScreen';
import { Icon } from 'react-native-elements';
import { OnBoarding } from './components/OnBoarding/onBoarding';

const headerStyles = StyleSheet.create({
  header: {
    backgroundColor: '#151515',
  },
  title: {
    color: "white"
  }
})

const AppRouter = () => {
  const Stack = createStackNavigator();
  const Tab = createBottomTabNavigator();

  function publicRoute() {
    return (
      // <NavigationContainer>
        <Tab.Navigator
          screenOptions={({ route }) => ({
            tabBarIcon: ({ focused, color, size }) => {
              let iconName;
  
              if (route.name === 'Home') {
                iconName = 'globe';
              } else if (route.name === 'Profile') {
                iconName = 'user';
              }
  
              return <Icon type="font-awesome" name={iconName} size={size} color={color} />;
            },
          })}
          tabBarOptions={{
            activeTintColor: 'red',
            inactiveTintColor: 'gray',
            style: {backgroundColor: '#4c4c4c'}
          }}
        >
          <Tab.Screen name='Home' component={HomeScreen} options={{ headerShown: false }} />
          <Tab.Screen name='Profile' component={ProfileScreen} options={{ headerShown: false }} />
        </Tab.Navigator> 
      // </NavigationContainer>
    )
  }

  function onBoardingRoute() {
    return (
      <NavigationContainer>
        <Stack.Navigator initialRouteName="OnBoarding">
          <Stack.Screen name="OnBoarding" component={OnBoarding} options={{ headerShown: false }} />
          <Stack.Screen name="Session" component={sessionRoute} options={{ headerShown: false }} />
          <Stack.Screen name="Public" component={publicRoute} options={{ headerShown: false }} />
        </Stack.Navigator>
      </NavigationContainer>
    )
  }

  function sessionRoute() {
    return (
      // <NavigationContainer>
        <Stack.Navigator>
          <Stack.Screen name = "Root" component={RootScreen} options={{
            headerShown: false 
          }} />

          <Stack.Screen name = "Login" component={LoginScreen} options={{
            headerShown: false,
          }} />

          <Stack.Screen name = "Register" component={RegisterScreen} options={{
            headerShown: false,
          }} />
        </Stack.Navigator>
      // </NavigationContainer>
    )
  }


  return onBoardingRoute();
}

export { AppRouter };
