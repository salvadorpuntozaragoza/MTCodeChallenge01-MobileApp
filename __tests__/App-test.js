/**
 * @format
 */

import 'react-native';
import React from 'react';
import { Provider } from 'react-redux';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { keyExtractor, validateEmail } from '../src/utils';
import { Root } from '../src/components/RootScreen/Root'
import { Login } from '../src/components/LoginScreen/Login';
import { Register } from '../src/components/RegisterScreen/Register';

test('Root screen renders correctly', () => {
  const tree = renderer.create(<Root />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Login screen renders correctly', () => {
  const tree = renderer.create(<Login />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Register screen renders correctly', () => {
  const tree = renderer.create(<Register />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Test that keyExtractor function returns an index as string', () => {
  expect(keyExtractor("", 2)).toBe("2");
});

test('Test that szaragoza@arkusnexus.com is a valid email', () => {
  expect(validateEmail("szaragoza@arkusnexus.com")).toBe(true);
});

test('Test that zaragoza.arkusnexu.com is an invalid email', () => {
  expect(validateEmail("szaragoza.arkusnexus.com,")).toBe(false);
});

// it('renders correctly', () => {
//   renderer.create(<App />);
// });
