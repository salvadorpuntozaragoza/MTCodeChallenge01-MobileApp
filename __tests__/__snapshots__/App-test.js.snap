// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root screen renders correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#1c1c1c",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "flex": 2,
        "height": 1334,
        "width": 750,
      }
    }
  >
    <Image
      height={60}
      resizeMode="center"
      source={
        Object {
          "testUri": "../../../src/assets/images/logo.png",
        }
      }
      width={60}
    />
  </View>
  <View
    style={
      Object {
        "alignSelf": "center",
        "flex": 1,
        "height": 400,
        "justifyContent": "center",
        "paddingHorizontal": 37.5,
        "width": 637.5,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 20,
          "overflow": "hidden",
          "shadowColor": "rgba(0,0,0, .4)",
          "shadowOffset": Object {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fe1546",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            Log in
          </Text>
        </View>
      </View>
    </View>
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#fef8f7",
          "fontSize": 12,
          "marginVertical": 10,
        }
      }
    >
      Or
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#fff",
          "borderRadius": 20,
          "overflow": "hidden",
          "shadowColor": "rgba(0,0,0, .4)",
          "shadowOffset": Object {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 1,
          "shadowRadius": 1,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fe1546",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            Sign up
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
